### DISCUSSION TITLE
Repl?

### DISCUSSION THREAD

**Discussion by ktodyruik** [qwertie/ecsharp#23] [Q&A]
Could loyc macros and enhanced c# be used in a Repl?

**Comment by qwertie** [comment]
I haven't looked into this. I think somebody wrote a C# repl ... let's see, [here's one from Mono](http://www.mono-project.com/docs/tools+libraries/tools/repl/). _Perhaps_ it could be adapted to run LeMP code.

**Comment by qwertie** [comment]
Actually I think Roslyn has a REPL engine (CSharpScriptEngine?), and it would be better to use that one. Since it's open source, it could be adapted to take LeMP as a front-end. A REPL, in turn, would be a great starting point for a EC# Visual Studio extension that provides IntelliSense for `.ecs` files. Plus, maybe the REPL itself could be integrated into VS. That is, if you're interested in doing that kind of work.

Maybe I misunderstood though, maybe all you want is a REPL that runs macros. If that's what you want, `LeMP.exe --editor` could be used as a substitute.

**Comment by qwertie** [comment]
Here's a fun idea: implement a C# interpreter... that could be a [fun thing to do with `matchCode`](http://loyc.net/lemp/ref-other.html#matchcode). Maybe I should write an article about doing that...

**Comment by qwertie** [comment]
Really though... better to just convert some code to Roslyn and then compile that

OMG, I know. I could write a `quoteRoslyn` macro that generates Roslyn syntax trees from EC# ones. Or I would if I had time. Btw I think I'll rename this repo to ecsharp.

**Comment by qwertie** [comment]
@jonathanvdc actually made a REPL called [LeMP-repl](https://github.com/jonathanvdc/LeMP-repl), but it hasn't been updated recently. However, there is a `LeMPDemo.exe` included in the zip file for the [release](https://github.com/qwertie/ecsharp/releases). It processes input on the left side to produce output on the right side. Be careful about not saving the file - invalid input sometimes crashes LeMP, so please save your work occasionally. And of course, you can use the `compileTime` and `precompute` macros to execute Enhanced C# code (otherwise the code is merely preprocessed).

If you would like to see the Loyc tree that corresponds to some Enhanced C# code, add `--outlang=les3` to the options list and click Regen.
